File: gitdot-auth/src/handler/device/create_device_code.rs
use axum::{Json, extract::State, http::StatusCode};

use gitdot_api::endpoint::auth::device::create_device_code as api;
use gitdot_core::dto::DeviceCodeRequest;

use crate::{
    app::{AppError, AppResponse, AppState},
    dto::IntoApi,
};

pub async fn create_device_code(
    State(state): State<AppState>,
    Json(body): Json<api::CreateDeviceCodeRequest>,
) -> Result<AppResponse<api::CreateDeviceCodeResponse>, AppError> {
    let request = DeviceCodeRequest {
        client_id: body.client_id,
        verification_url: state.settings.gitdot_oauth_device_verification_url.clone(),
    };
    state
        .device_service
        .request_device_code(request)
        .await
        .map_err(AppError::from)
        .map(|code| AppResponse::new(StatusCode::CREATED, code.into_api()))
}

File: gitdot-web/packages/gitdot-api/src/endpoint/auth/device/create-device-code.ts
import { z } from "zod";
import { DeviceCodeResource } from "../../../resource";
import type { Endpoint } from "../../endpoint";

export const CreateDeviceCodeRequest = z.object({
  client_id: z.string(),
});
export type CreateDeviceCodeRequest = z.infer<typeof CreateDeviceCodeRequest>;

export const CreateDeviceCodeResponse = DeviceCodeResource;
export type CreateDeviceCodeResponse = z.infer<typeof CreateDeviceCodeResponse>;

export const CreateDeviceCode = {
  path: "/auth/device/code",
  method: "POST",
  request: CreateDeviceCodeRequest,
  response: CreateDeviceCodeResponse,
} as const satisfies Endpoint;

export type CreateDeviceCode = typeof CreateDeviceCode;
